File: 67-add-binary/67-add-binary.scala
object Solution {
    def addBinary(a: String, b: String): String = {
        var i = a.length - 1
        var j = b.length - 1
        var res = ""
        var c = 0
        
        while (i >= 0 || j >= 0) {
            val v1 = if (i >= 0 && a.charAt(i) == '1') 1 else 0
            val v2 = if (j >= 0 && b.charAt(j) == '1') 1 else 0
            val s = v1 + v2 + c
            
            c = if (s >= 2) 1 else 0
            
            res += (s % 2)
            
            i -= 1
            j -= 1
        }
        
        if (c > 0) res += "1"
        
        res.reverse
    }
}

File: 67-add-binary/67-add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        int i1 = a.length() - 1;
        int i2 = b.length() - 1;
        int c = 0;
        
        StringBuilder result = new StringBuilder();
        
        while(i1 >= 0 || i2 >= 0) {
            int v1 = (i1 >= 0)? Character.getNumericValue(a.charAt(i1--)) : 0;
            int v2 = (i2 >= 0)? Character.getNumericValue(b.charAt(i2--)) : 0;
            
            int s = v1 + v2 + c;
            result.append((s % 2));
            c = s / 2;
        }
        
        if (c != 0) result.append(1);
        
        result = result.reverse();
        
        return result.toString();
    }
}